src/uucore/src/lib/features/perms.rs

```diff
@@ -188,6 +189,77 @@ pub struct ChownExecutor {
     pub dereference: bool,
 }
 
+#[cfg(test)]
+pub fn check_root(path: &Path, would_recurse_symlink: bool) -> bool {
+    is_root(path, would_recurse_symlink)
+}
+
+fn is_root(path: &Path, would_recurse_symlink: bool) -> bool {
+    // Prohibit only if:
+    // (--preserve-root and -R present) &&
+    // (path canonicalizes to "/") &&
+    // (
+    //     (path is a symlink && would recurse on this symlink) ||
+    //     (path is not a symlink)
+    // )
+    // The first clause is checked by the caller, the second and third clause is checked here.
+    // The caller has to evaluate -P/-H/-L into 'would_recurse_symlink'.
+    // Recall that canonicalization resolves both relative paths (e.g. "..") and symlinks.
+
+    // The third clause can be evaluated without any syscalls, so we do that first.
+    // If we would_recurse_symlink, then the clause is true no matter whether the path is a symlink
+    // or not. Otherwise, we only need to check here if the path can syntactically be a symlink:
+    if !would_recurse_symlink {
+        // We cannot check path.is_dir() here, as this would resolve symlinks,
+        // which we need to avoid here.
+        // All directory-ish paths match "*/", except ".", "..", "*/.", and "*/..".
+        let looks_like_dir = match path.as_os_str().to_str() {
+            // If it contains special character, prefer to err on the side of safety, i.e. forbidding the chown operation:
+            None => false,
+            Some(".") | Some("..") => true,
+            Some(path_str) => {
+                (path_str.ends_with(MAIN_SEPARATOR_STR))
+                    || (path_str.ends_with(&format!("{}.", MAIN_SEPARATOR_STR)))
+                    || (path_str.ends_with(&format!("{}..", MAIN_SEPARATOR_STR)))
+            }
+        };
+        // TODO: Once we reach MSRV 1.74.0, replace this abomination by something simpler, e.g. this:
+        // let path_bytes = path.as_os_str().as_encoded_bytes();
+        // let looks_like_dir = path_bytes == [b'.']
+        //     || path_bytes == [b'.', b'.']
+        //     || path_bytes.ends_with(&[MAIN_SEPARATOR as u8])
+        //     || path_bytes.ends_with(&[MAIN_SEPARATOR as u8, b'.'])
+        //     || path_bytes.ends_with(&[MAIN_SEPARATOR as u8, b'.', b'.']);
+        if !looks_like_dir {
+            return false;
+        }
+    }
+
+    // FIXME: TOCTOU bug! canonicalize() runs at a different time than WalkDir's recursion decision.
+    // However, we're forced to make the decision whether to warn about --preserve-root
+    // *before* even attempting to chown the path, let alone doing the stat inside WalkDir.
+    if let Ok(p) = path.canonicalize() {
+        let path_buf = path.to_path_buf();
+        if p.parent().is_none() {
+            match path_buf.to_str() {
```

Comment:
We can print osstrs with quotes using the Quotable trait: https://docs.rs/uucore/latest/uucore/display/trait.Quotable.html

Reply:
The GNU tests check for specific error messages, and this match statement just accesses the internal `&OsStr` to check for a specific value. What do you mean, which improvement did I miss?

Reply:
Yes, the messsage is correct, but I think the conversion to a `str` can be avoided with that quotable trait. It can print `OsStr` directly. But maybe there's something I'm missing?

Reply:
Ah! I think I see what you mean. Yes, I'll address it.